### PR TITLE
Fix #25552: Add Cache Implementation to retrieve License Info Once

### LINKED ISSUE #25552: Loading the /api/v1/appconfiguration every mouse click
### Parent Issue

_No response_

### Problem Statement

We seem to be loading the `/api/v1/appconfiguration` endpoint with every mouse click on our main navigation.  Take a look at the screencap.  I tried it with "Disable Cache" checked and unchecked and saw the same behavior.

### Steps to Reproduce

Open your Network tab and browse the backend, you will see repeated calls to `/api/v1/appconfiguration`

https://github.com/dotCMS/core/assets/934364/237c658e-a665-424a-9bde-8914e052a66c

### Acceptance Criteria

There should be 1 call to the `/api/v1/appconfiguration` and the result should be stored in our app for future use and never requested again.

### dotCMS Version

23.07

### Proposed Objective

Application Performance

### Proposed Priority

Please Select

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.

_No response_

### Assumptions & Initiation Needs

_No response_

### Quality Assurance Notes & Workarounds

_No response_

### Sub-Tasks & Estimates

_No response_

### PR DESCRIPTION
**Pull request by zJaaal**
### Summary
🧪🛠️🚀

The pull request adds and updates the logic for handling the license data in the `AppComponent` and the `DotLicenseService`. It also adds test cases for these components and refactors the `DotcmsConfigService` file.

> _Oh we're the coders of the `DotLicenseService`_
> _We cache the data and we test it well_
> _We heave and ho and we refactor the code_
> _And we make the `AppComponent` sail_

### Walkthrough
*  Add `DotLicenseService` to `AppComponent` to handle license data ([link](https://github.com/dotCMS/core/pull/25646/files?diff=unified&w=0#diff-c0762a9384069a32c029261a27d2ab7ae8bce5eea7850d8188a93f0cd8502341L12-R12), [link](https://github.com/dotCMS/core/pull/25646/files?diff=unified&w=0#diff-c0762a9384069a32c029261a27d2ab7ae8bce5eea7850d8188a93f0cd8502341R30), [link](https://github.com/dotCMS/core/pull/25646/files?diff=unified&w=0#diff-c0762a9384069a32c029261a27d2ab7ae8bce5eea7850d8188a93f0cd8502341L41-R43), [link](https://github.com/dotCMS/core/pull/25646/files?diff=unified&w=0#diff-c0762a9384069a32c029261a27d2ab7ae8bce5eea7850d8188a93f0cd8502341R53), [link](https://github.com/dotCMS/core/pull/25646/files?diff=unified&w=0#diff-65898cd088598c46bc62504b0033ddfe90debf41720d18cbb892ed35f857598eL6-R8), [link](https://github.com/dotCMS/core/pull/25646/files?diff=unified&w=0#diff-65898cd088598c46bc62504b0033ddfe90debf41720d18cbb892ed35f857598eL21-R23), [link](https://github.com/dotCMS/core/pull/25646/files?diff=unified&w=0#diff-65898cd088598c46bc62504b0033ddfe90debf41720d18cbb892ed35f857598eL33-R36), [link](https://github.com/dotCMS/core/pull/25646/files?diff=unified&w=0#diff-65898cd088598c46bc62504b0033ddfe90debf41720d18cbb892ed35f857598eL41-R55))
*  Add `license` property to mock `configParams` object in `AppComponent` spec file to simulate license data from `DotcmsConfigService` ([link](https://github.com/dotCMS/core/pull/25646/files?diff=unified&w=0#diff-c0762a9384069a32c029261a27d2ab7ae8bce5eea7850d8188a93f0cd8502341R64-R69))
*  Add spy on `setLicense` method of `dotLicenseService` in `AppComponent` spec file to verify its invocation ([link](https://github.com/dotCMS/core/pull/25646/files?diff=unified&w=0#diff-c0762a9384069a32c029261a27d2ab7ae8bce5eea7850d8188a93f0cd8502341R75))
*  Add test case to check that `setLicense` method of `dotLicenseService` is called with expected license data in `AppComponent` spec file ([link](https://github.com/dotCMS/core/pull/25646/files?diff=unified&w=0#diff-c0762a9384069a32c029261a27d2ab7ae8bce5eea7850d8188a93f0cd8502341R96-R104))
*  Add `license` property to `DotLicenseService` class to store cached license data ([link](https://github.com/dotCMS/core/pull/25646/files?diff=unified&w=0#diff-6e20680ebb3defa4a38ced2e3d445ed08618df5978c87e26e7808faa8d989ec7R83-R84))
*  Add `setLicense` method to `DotLicenseService` class to set cached license data ([link](https://github.com/dotCMS/core/pull/25646/files?diff=unified&w=0#diff-6e20680ebb3defa4a38ced2e3d445ed08618df5978c87e26e7808faa8d989ec7L127-R151))
*  Modify `getLicense` method of `DotLicenseService` class to return cached license data if it exists, or to make a request and cache the license data using `setLicense` method ([link](https://github.com/dotCMS/core/pull/25646/files?diff=unified&w=0#diff-6e20680ebb3defa4a38ced2e3d445ed08618df5978c87e26e7808faa8d989ec7L127-R151))
*  Add test case to check that `getLicense` method of `DotLicenseService` does not make any request if license is already set by `setLicense` method in `DotLicenseService` spec file ([link](https://github.com/dotCMS/core/pull/25646/files?diff=unified&w=0#diff-786c97306607eae52dcbde48b669c5becc4e99e32f6db3840e11b4db5e791512R138-R152))
*  Reorder imports alphabetically and add blank lines for readability in `DotcmsConfigService` file ([link](https://github.com/dotCMS/core/pull/25646/files?diff=unified&w=0#diff-9d02a07b0759480653f520aaeb3df576768ab3d445afbff8db9da1a24eb0e800L1-R7), [link](https://github.com/dotCMS/core/pull/25646/files?diff=unified&w=0#diff-9d02a07b0759480653f520aaeb3df576768ab3d445afbff8db9da1a24eb0e800L137-R144))

### Screenshots

https://github.com/dotCMS/core/assets/63567962/7d668b1b-11e0-42eb-9588-937c17cbbb0d

### CODE REVIEW

**Review by danielmdob — APPROVED** [review]

File: core-web/libs/data-access/src/lib/dot-license/dot-license.service.ts [hunk 1]
```diff
@@ -124,10 +126,43 @@ export class DotLicenseService {
...
+                })
+            );
+    }
+
+    /**
+     * Update cached license
+     *
+     * @memberof DotLicenseService
+     */
+    public updateLicense(): void {
```
Comment (fmontes) [inline core-web/libs/data-access/src/lib/dot-license/dot-license.service.ts:148]:
```suggestion
    updateLicense(): void {
```

File: core-web/libs/data-access/src/lib/dot-license/dot-license.service.ts [hunk 2]
```diff
@@ -124,10 +126,43 @@ export class DotLicenseService {
...
     private getLicense(): Observable<DotLicense> {
+        if (this.license) return of(this.license);
+
         return this.coreWebService
             .requestView({
                 url: this.licenseURL
             })
-            .pipe(pluck('entity', 'config', 'license'));
+            .pipe(
+                pluck('entity', 'config', 'license'),
```
Comment (fmontes) [inline core-web/libs/data-access/src/lib/dot-license/dot-license.service.ts:136]: Do we need a `take(1)`?

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/libs/data-access/src/lib/dot-license/dot-license.service.ts (hunk 2 above)
Comment (fmontes) [inline core-web/libs/data-access/src/lib/dot-license/dot-license.service.ts:136]: Do we need a take(1)?

File: core-web/libs/data-access/src/lib/dot-license/dot-license.service.ts (hunk 1 above)
Comment (fmontes) [inline core-web/libs/data-access/src/lib/dot-license/dot-license.service.ts:148]:
```suggestion
    updateLicense(): void {
```

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/libs/data-access/src/lib/dot-license/dot-license.service.ts (hunk 2 above)
Comment (zJaaal) [reply core-web/libs/data-access/src/lib/dot-license/dot-license.service.ts:136]: It's not needed, `coreWebService` uses `HttpClient` to fetch and that service uses the `take(1)` under the hood

**Review by fmontes — APPROVED** [review]

**Pull request merged**